R/utils-code_dependency.R

```diff
@@ -0,0 +1,337 @@
+    parsed_code <- parse(text = code)
+  }
+
+  pd <- utils::getParseData(parsed_code)
+
+  calls_pd <- lapply(pd[pd$parent == 0, "id"], get_children, pd = pd)
+
+  occurrence <- lapply(lapply(object_names, detect_symbol, pd = calls_pd), which)
+
+  names(occurrence) <- object_names
+
+  cooccurrence <- lapply(
+    calls_pd,
+    function(x) {
+      sym_cond <- which(x$token %in% c("SYMBOL", "SYMBOL_FUNCTION_CALL") & x$text %in% object_names)
+      sym_form_cond <- which(x$token == "SYMBOL_FORMALS" & x$text %in% object_names)
+      sym_cond <- sym_cond[!x[sym_cond, "text"] %in% x[sym_form_cond, "text"]]
+
+      if (length(sym_cond) >= 2) {
+        ass_cond <- grep("ASSIGN", x$token)
+        text <- unique(x[sort(c(sym_cond, ass_cond)), "text"])
+
+        if (text[1] == "->") {
+          rev(text[-1])
+        } else {
+          text[-1]
+        }
+      }
+    }
+  )
+
+  side_effects <- grep("@effect", pd[pd$token == "COMMENT", "text"], value = TRUE)
+  check_effects <-
+    if (length(side_effects) > 0) {
+      affected <-
+        unlist(strsplit(sub("\\s*#\\s*@effect\\s+", "", side_effects), "\\s+"))
+
+      unique(c(object_names, affected))
+    } else {
+      object_names
+    }
+
+  effects <- lapply(
+    check_effects,
+    function(x) {
+      maxid <- max(occurrence[[x]])
+      return_code_for_effects(
+        x,
+        pd = calls_pd,
+        occur = suppressWarnings(lapply(occurrence, function(x) setdiff(x, maxid:max(maxid, max(x))))),
+        cooccur = cooccurrence,
+        eff = NULL
+      )
+    }
+  )
+  names(effects) <- check_effects
+
+  list(
+    occurrence = occurrence,
+    cooccurrence = cooccurrence,
+    effects = effects
+  )
+}
+
+#' @title Get children calls within `getParseData()` object
+#' @param pd `list` of `data.frame`s of results of `utils::getParseData()` trimmed to unique `parsed_code` calls
+#' @param parent parent id in `utils::getParseData()`
+#' @return Row `binded` `utils::getParseData()` of all calls.
+#' @keywords internal
+get_children <- function(pd, parent) {
+  idx_children <- abs(pd$parent) == parent
+  children <- pd[idx_children, c("token", "text", "id")]
+  if (nrow(children) == 0) {
+    return(NULL)
+  }
+
+  if (parent > 0) {
+    do.call(rbind, c(list(children), lapply(children$id, get_children, pd = pd)))
+  }
+}
+
+#' @title Detects `"SYMBOL"` tokens for row `binded` `getParseData()` structure
+#' @param object `character` containing the name of the object
+#' @param pd `list` of `data.frame`s of results of `utils::getParseData()` trimmed to unique `parsed_code` calls
+#' @return A `logical` vector pointing in which elements of `pd` the `SYMBOL` token row has `object` in text column
+#' @keywords internal
+detect_symbol <- function(object, pd) {
+  unlist(
+    vapply(
+      pd,
+      function(call) {
+        any(call[call$token %in% c("SYMBOL", "SYMBOL_FUNCTION_CALL"), "text"] == object) &&
+          !any(call[call$token == "SYMBOL_FORMALS", "text"] == object)
```

Comment:
This means that function (symbol) can't be named in the same way as one of the formals. I don't see it is very probable and not easy to technicaly exclude this case. Please failing test with skip (I don't know if CRAN accepts skip)

```r
testthat::test_that("Can't detect occurrence of function definition when a formal is named as a function", {
  q <- new_qenv()
  q <- eval_code(q, "x <- 1")
  q <- eval_code(q, "foo <- function(foo = 1) 'elo'")
  q <- eval_code(q, "a <- foo(x)")

  testthat::expect_identical(
    get_code(q, names = "a"),
    c("x <- 1", "a <- foo(x)")
  )
})
```

Reply:
I think I could work it out, currently this is as below
> If SYMBOL exists on LHS or on RHS but **any** occurence is a SYMBOL_FORMALS then it is not occurence
but we could change it to
> If SYMBOL exists on LHS or on RHS but **all** occurences are SYMBOL_FORMALS than it is not occurence

Reply:
I fixed this situation in here 0e5e6ea6445376d174a082b85478bf611997225a